package PS;
import "util/proto/range.proto";
import "data/proto/data.proto";
import "system/proto/node.proto";
import "parameter/proto/param.proto";
import "filter/proto/filter.proto";
import "learner/proto/sgd.proto";
import "learner/proto/bcd.proto";

message Task {
  // true: system control task, typically *ctrl* should be set
  // false: a task for a customer, and *customer_id* should be set
  optional bool control = 1 [default = false];
  // true: a request task
  // false: the response task to the request task with the same *time*
  optional bool request = 2 [default = false];
  // the unique id of a customer
  optional int32 customer_id = 3;

  // the timestamp if this task
  optional int32 time = 5;
  // the depended tasks of this one. that is, this task is executed only if all
  // tasks from the same node with time contained in *wait_time* are finished.
  // only valid if *request*=true
  repeated int32 wait_time = 6;

  // the key range of this task
  optional PbRange key_range = 7;
  // namespace of keys
  optional int32 key_channel = 8;
  // true: the message sent with this task will contain a list of keys
  optional bool has_key = 9 [default = false];

  // data type
  optional DataType key_type = 13;
  repeated DataType value_type = 14;

  // filters applied to the data
  repeated FilterConfig filter = 12;

  // if true, the tasks will be packed into a single one during sending
  optional bool more = 16 [default = false];
  repeated Task task = 15;

  // the place to store a small amount of data
  optional bytes msg = 17;

  // system control signals
  optional Control ctrl = 18;

  // parameters
  optional ParamCall param = 20;
  optional SGDCall sgd = 21;
  optional BCDCall bcd = 22;

  extensions 100 to 199;
}

message Control {
  enum Command {
    // a node => the scheduler
    REQUEST_APP = 1;
    REGISTER_NODE = 2;
    REPORT_PERF = 3;
    READY_TO_EXIT = 4;

    // the scheduler => a node
    ADD_NODE = 10;
    UPDATE_NODE = 11;
    REPLACE_NODE = 12;
    REMOVE_NODE = 13;
    EXIT = 14;
  }
  required Command cmd = 1;
  repeated Node node = 2;
}

enum DataType {
  OTHER = 0;
  INT8 = 1;
  INT16 = 2;
  INT32 = 3;
  INT64 = 4;
  UINT8 = 5;
  UINT16 = 6;
  UINT32 = 7;
  UINT64 = 8;
  FLOAT = 9;
  DOUBLE = 10;
  CHAR = 11;
}
